### DISCUSSION TITLE
PySR for Wave Identification

### DISCUSSION THREAD

**Discussion by rodrigochi** [MilesCranmer/PySR#552] [Q&A]
Hello all,

I've been working with PySR in an attempt to detect traveling waves within noisy data sets, specifically aiming to model the data using expressions such as  $`Ae^{-\alpha t} \cos(\omega*t - k*t)`$ (allowing also constant terms). Despite some progress, the results have not fully met my expectations. The outcomes are close, but I believe there's potential for improvement.

A recurring issue is the algorithm's tendency to overfit the data, often resulting in a final expression that deviates from what seems to be an initially more accurate representation. This challenge has made it difficult to accurately capture the dynamics I'm interested in.

I'm here seeking advice or suggestions on how to refine my approach to achieve better fitting results and reduce the overfitting problem. Below, I've shared a snippet of my code for context. Any feedback on how to improve this process or alternative methods to consider would be greatly appreciated.

Best,
Rodrigo

```python
import numpy as np
from pysr import PySRRegressor
# Generate synthetic data
X = np.hstack([np.linspace(-1, 1, 100).reshape(-1, 1), np.linspace(0, 10, 100).reshape(-1, 1)]) # [Space, Time]
y = 2.5 * np.cos(3* X[:, 0] - 2 * X[:, 1]) * np.exp(-0.1 * X[:, 1]) - np.sin(X[:, 0] + 3 * X[:, 1]) * np.exp(-0.3 * X[:, 1]) #+ np.random.normal(0,1,1000).reshape(-1, 1) 
# Configure PySRRegressor with nested constraints
model = PySRRegressor(
    procs=10,
    populations=20,
    niterations=200,  # Adjust the number of iterations as needed
    binary_operators=["+", "*"],
    unary_operators=["exp", "sin", "cos"],
    nested_constraints={
        "sin": {"sin": 0, "cos": 0, "exp": 0, '+':1},
        "cos": {"sin": 0, "cos": 0, "exp": 0, '+':1},
        "exp": {"sin": 0, "cos": 0, "exp": 0, '+':0},
    },
)
# Fit the model
model.fit(X, y)

# Print the best equation found
print("Best equation:", model)
```

**Comment by MilesCranmer** [comment]
Welcome @rodrigochi!

Have you tried the tuning tips on https://astroautomata.com/PySR/tuning/? I would start with those. For example, you should not use both `sin` and `cos`, because you can build one out of the other (since you have a `+` available).
